gr-picoscope/picoscope.hpp

```diff
@@ -0,0 +1,559 @@
+    std::string        serial_number;
+    double             sample_rate              = 10000.;
+    acquisition_mode_t acquisition_mode         = acquisition_mode_t::STREAMING;
+    std::size_t        pre_samples              = 1000;
+    std::size_t        post_samples             = 9000;
+    std::size_t        rapid_block_nr_captures  = 1;
+    bool               trigger_once             = false;
+    double             streaming_mode_poll_rate = 0.001;
+    bool               auto_arm                 = true;
+    ChannelMap         enabled_channels;
+    PortMap            enabled_ports;
+    trigger_setting_t  trigger;
+};
+
+struct State {
+    std::vector<Channel>        channels;
+    std::atomic<std::size_t>    data_available     = 0;
+    std::atomic<bool>           data_finished      = false;
+    bool                        initialized        = false;
+    bool                        configured         = false;
+    bool                        closed             = false;
+    bool                        armed              = false;
+    bool                        started            = false; // TODO transitional until nodes have state handling
+    int16_t                     handle             = -1;    ///< picoscope handle
+    int16_t                     overflow           = 0;     ///< status returned from getValues
+    int16_t                     max_value          = 0;     ///< maximum ADC count used for ADC conversion
+    double                      actual_sample_rate = 0;
+    std::thread                 poller;
+    BufferHelper<Error>         errors;
+    std::atomic<poller_state_t> poller_state = poller_state_t::IDLE;
+    std::atomic<bool>           forced_quit  = false; // TODO transitional until we found out what
+                                                      // goes wrong with multithreaded scheduler
+    std::size_t produced_worker = 0;                  // poller/callback thread
+};
+
+} // namespace detail
+
+// optional shortening
+template<typename T, fair::meta::fixed_string description = "", typename... Arguments>
+using A = fair::graph::Annotated<T, description, Arguments...>;
+
+using fair::graph::Visible;
+
+template<typename PSImpl>
+struct Picoscope : public fair::graph::node<PSImpl, fair::graph::BlockingIO<true>> {
+    A<std::string, "serial number", Visible> serial_number;
+    A<double, "sample rate", Visible>        sample_rate = 10000.;
+    // TODO any way to get custom enums into pmtv??
+    A<std::string, "acquisition mode as string", Visible>      acquisition_mode_string  = std::string("STREAMING");
+    A<std::size_t, "pre-samples", Visible>                     pre_samples              = 1000;
+    A<std::size_t, "post-samples", Visible>                    post_samples             = 9000;
+    A<std::size_t, "no. captures (rapid block mode)", Visible> rapid_block_nr_captures  = 1;
+    A<bool, "trigger once (rapid block mode)", Visible>        trigger_once             = false;
```

Comment:
Need to check which fields need to be actually declared 'Visible'. Ideally only the most important 4-5 ... maybe in this case more but having all Visible makes an impossible block UI visualisation in the flow-graph.

Reply:
Ah, didn't know that this was about UI visibility. Reduced this to sample_rate and acquisition_mode now.

Reply:
Removed the ctor. The dtor currently ensures that the driver is disconnected from the device, so that's kind of important...